### PR TITLE
Prevent dialog asking to enter dangerous terrain

### LINKED ISSUE #28215: Feature/fix: Moving over this repeated message is slow.
**Is your feature request related to a problem? Please describe.**  
There are several majorly spammy or even game-interrupting messages that detract from basic gameplay elements while moving. The two key offenders seem to be:

"Moving over this long grass is slow"
and
"Do you really want to step into that blackberry bush, Y/N"

Edit: anything related to jumping into or diving out of deep water is also a major culprit.

**Describe the solution you'd like**  
1. An options setting to silence identical movement related error messages for X minutes, where X is specified by the user and 0 minutes would maintain the current behaviour. 
- So, if set to 5 minutes, then once per 5 minutes you would get a "moving over this long grass is slow" message when getting onto long grass. You might still get a "moving over this giant pile of moose turds is slow" message while the long grass one was silenced, because they're different messages.
- I would suggest keeping the hidden messages visible in the message log though. Only hide them in the sidebar.

2. Make harmful terrain impassable while walking (including pits, identified traps. and things as well as rose bushes), and:
- Replace the "do you really want to step into the hip-deep pool of molten lead Y/N" prompts with a message in the message log saying "Stepping into that hip-deep pool of molten lead looks risky.  Press (run key) and move into it to force movement."
- Allow movement message silencing in options to apply to the "stepping into" message as well

Making "run" the way to force movement into hazards would improve the feeling of stepping into hazards anyway, making it harder for players to step into a deep pit of spikes when travelling at walking speed (something that should be difficult) but allowing them to do it accidentally when fleeing a zombear (something that should be easy to do).

### PR DESCRIPTION
**Pull request by Ramza13**
#### Summary
SUMMARY: Interface "Prevent dialog asking to enter dangerous terrain"

#### Purpose of change
Fixes #28215
When attempting to walk into dangerous terrain it will be treated as a wall and print a message telling you to run if you want to enter it.  If you run into it you will go onto it with no warning or you will be warned depending on a new option which defaults to warning you.

### CODE REVIEW

File: src/game.cpp [hunk 1]
```diff
@@ -5648,6 +5648,11 @@ void game::examine( const tripoint &examp )
             Pickup::pick_up( examp, 0 );
         }
     }
+    if( is_dangerous_tile( examp ) ) {
+        if( prompt_dangerous_tile( examp ) ) {
```
Comment (Night-Pryanik) [inline src/game.cpp]: Could be merged into one `if( is_dangerous_tile( examp ) && prompt_dangerous_tile( examp ) )` check.

Comment (Ramza13) [reply src/game.cpp]: Good catch.

**Comment by Ramza13** [comment]
I just looked again at the parent PR and saw that the idea was to have running just take you onto dangerous terrain without warning.  I have implemented that and for now left the examine option as it seems a more easily discoverable way to move onto dangerous stuff but can see an argument for removing it.

**Comment by Ramza13** [comment]
Took another look at parent PR and realized the issue was solved there.  Removed examine logic and printed message to tell user to run if they want to move onto terrain.

File: src/game.cpp [hunk 2]
```diff
@@ -9144,7 +9144,14 @@ bool game::walk_move( const tripoint &dest_loc )
     }
     u.set_underwater( false );
 
-    if( !shifting_furniture && !pushing && !prompt_dangerous_tile( dest_loc ) ) {
+    if( !shifting_furniture && !pushing && ( is_dangerous_tile( dest_loc ) &&
+            !u.movement_mode_is( CMM_RUN ) ) ) {
+        if( is_dangerous_tile( dest_loc ) && !u.movement_mode_is( CMM_RUN ) ) {
```
Comment (BevapDin) [inline src/game.cpp]: This was already checked in the lines above. It won't have changed. You don't need to check it again.

Comment (Ramza13) [reply src/game.cpp]: Good catch, thanks

**Comment by I-am-Erk** [comment]
Glad to finally see this being implemented!

**Comment by Night-Pryanik** [comment]
While the rationale might be good, what if I want to walk or crouch into dangerous terrain?

**Comment by I-am-Erk** [comment]
Hit run, move one tile. The number of cases when it's going to be something you really want to do are vanishingly small.

**Comment by Ramza13** [comment]
> 
> 
> While the rationale might be good, what if I want to walk or crouch into dangerous terrain?

At one point in this I had added to the explore option the option to move into the dangerous terrain.  That solved your question but made it slightly more annoying to examine dangerous terrain.  Should I add that back in?

**Comment by I-am-Erk** [comment]
Since many dangerous terrains are berry shrubs, it would be very annoying to get a new dialogue every time you tried to pick berries

File: src/game.cpp [hunk 3]
```diff
@@ -9144,7 +9144,13 @@ bool game::walk_move( const tripoint &dest_loc )
     }
     u.set_underwater( false );
 
-    if( !shifting_furniture && !pushing && !prompt_dangerous_tile( dest_loc ) ) {
+    if( !shifting_furniture && !pushing && is_dangerous_tile( dest_loc ) &&
+        !u.movement_mode_is( CMM_RUN ) ) {
+        std::vector<std::string> harmful_stuff = get_dangerous_tile( dest_loc );
+        add_msg( m_warning,
+                 _( "Stepping into that %s looks risky. Press (run key) and move into it to force movement." ),
+                 enumerate_as_string( harmful_stuff ) );
```
Comment (Night-Pryanik) [inline src/game.cpp]: Replace with `_( "Stepping into that %1$s looks risky. %2$s and move into it to force movement." ),
                 enumerate_as_string( harmful_stuff ), press_x( ACTION_TOGGLE_RUN ) );`

**Comment by Fris0uman** [comment]
Does that mean that if you're runing away from something and not paying attention you will actually jum from roof/step in acid/Walk on traps ?

**Comment by Ramza13** [comment]
> 
> 
> Does that mean that if you're runing away from something and not paying attention you will actually jum from roof/step in acid/Walk on traps without any prompt?

Yes

**Comment by Zireael07** [comment]
That's not a good solution, then.

**Comment by Ramza13** [comment]
> That's not a good solution, then.

The idea, which came from #28215 is to make running a bit more dangerous if you aren't paying attention, sort of like how you can drive into things if you aren't paying attention.  If that solution doesn't appeal to you, how instead would you want to allow entering dangerous terrain?  A new key maybe or add it to examine again with a caveat about letting you pick berries first or something?

File: src/game.cpp [hunk 4]
```diff
@@ -9144,7 +9144,17 @@ bool game::walk_move( const tripoint &dest_loc )
     }
     u.set_underwater( false );
 
-    if( !shifting_furniture && !pushing && !prompt_dangerous_tile( dest_loc ) ) {
+    if( !shifting_furniture && !pushing && is_dangerous_tile( dest_loc ) &&
+        !u.movement_mode_is( CMM_RUN ) ) {
+        std::vector<std::string> harmful_stuff = get_dangerous_tile( dest_loc );
+        std::string key_message = press_x( ACTION_TOGGLE_RUN );
+        if( key_message == ( "Try" ) ) {
```
Comment (BevapDin) [inline src/game.cpp]: `press_x` returns a translated string. You are comparing this with a fixed string. That won't work when the user has a non-English language set.

**Comment by Leland** [comment]
Just chiming in to say that I also have reservations with this. 

People can make mistakes sometimes with inputs. Punishing that by death in a roguelike is very harsh.

And also this behavior doesn't ever actually seem to be communicated with the player. It's not intuitive either; I haven't seen this in another game and wouldn't expect it.

**Comment by I-am-Erk** [comment]
You've never seen a roguelike where you can move into hazardous terrain without a prompt? Really?

**Comment by esotericist** [comment]
While I am not opposed to this behavior as an option, I would really prefer it be an *option*. I rely heavily on safe mode going "are you sure?" to help me keep from going somewhere I did not intend that would have consequences. Taking that away while running would be a massive reduction in quality of life for me.

**Comment by Leland** [comment]
> You've never seen a game where you can move into hazardous terrain without a prompt?

I don't think most players will assume "toggle run mode" will also mean "disable warning prompts"

**Comment by Ramza13** [comment]
I added an option which defaults to false to prevent being warned while running.  If you have this option set to true you will not be warned while running.  If you have it set to false then when you run into a dangerous tile you will be warned.

File: src/options.cpp [hunk 5]
```diff
@@ -1139,6 +1139,11 @@ void options_manager::add_options_general()
          false
        );
 
+    add( "DANGEROUS_RUNNING", "general", translate_marker( "Dangerous running" ),
+         translate_marker( "If true, will not prevent the user from walking unto know hazardous tiles while running." ),
```
Comment (esotericist) [inline src/options.cpp]:
```suggestion
         translate_marker( "If true, the player will not be prevented from moving into known hazardous tiles while running." ),
```

File: src/game.cpp [hunk 6]
```diff
@@ -9144,8 +9144,20 @@ bool game::walk_move( const tripoint &dest_loc )
...
 
-    if( !shifting_furniture && !pushing && !prompt_dangerous_tile( dest_loc ) ) {
-        return true;
+
+    if( !shifting_furniture && !pushing && is_dangerous_tile( dest_loc ) ) {
+        if( !u.movement_mode_is( CMM_RUN ) ) {
+            std::vector<std::string> harmful_stuff = get_dangerous_tile( dest_loc );
+            std::string key_message = press_x( ACTION_TOGGLE_RUN );
+            add_msg( m_warning,
+                     _( "Stepping into that %1$s looks risky. %2$s to run into it to force movement." ),
```
Comment (esotericist) [inline src/game.cpp]: The verbiage in this warning is only correct if `DANGEROUS_RUNNING` is false, but you don't check for it here.

Further, `ACTION_TOGGLE_RUN` might not be defined for a player.

**Comment by esotericist** [comment]
I'm terribly sorry for how long it took for me to give feedback on this; I somehow completely spaced that you had put work into making the changes I was requesting

File: src/game.cpp (hunk 6 above)
Comment (esotericist) [reply src/game.cpp]: Given the number of ways someone might have their bindings set up for running (there's three completely different ways to set movement mode, and only cycling is bound by default), displaying a key here is probably not the way to go.

**Comment by Ramza13** [comment]
> 
> 
> I'm terribly sorry for how long it took for me to give feedback on this; I somehow completely spaced that you had put work into making the changes I was requesting

Ha no worries its the holiday season after all.  Thanks for helping me out!

**Comment by esotericist** [comment]
I started to complain that you hadn't addressed my concern about not checking for the config setting when moving into the tile, but I *only now* realized that turning on running still gets you the "are you sure?" prompt if dangerous running is off, so the instructions are correct.

I feel dumb for missing that, and I should have done actual testing.

Let me do that now.

**Comment by esotericist** [comment]
@Ramza13 The commit history for this is unnecessarily long considering how small a change it is, so I do need to urge you once again to learn more about how git works, and in particular how to set up branches off of upstream master (so you aren't carrying around all of that history) and how to rebase things when needed (to move changes to a newer version of upstream master).

Please try to work on this for other contributions?

**Comment by RDru** [comment]
I'm sorry but it's kinda illogical to have to run to be able to pass dangerous terrain. In reality you would prefer to walk carefully over them.

**Comment by I-am-Erk** [comment]
Wreckage is a bit of a niche scenario there, it's the only dangerous terrain I can think of where you might actually want to walk on it. We might need a flag for terrain like that to change the behaviour.

**Comment by esotericist** [comment]
> Wreckage is a bit of a niche scenario there, it's the only dangerous terrain I can think of where you might actually want to walk on it. We might need a flag for terrain like that to change the behaviour. Possibly making it climbable would work, then you'd be back to one press.

My hope is to actually get us a "safety" concept component to movement mode, alongside supported. That would cover this problem space (among others) much better than this PR did, but this does the job for now.

**Comment by esotericist** [comment]
> Plus this is worse than before. Even with the option set to true, Now I have to press cycle move 3 times to pass dangerous terrain when before I had to press once ('y').

Fwiw, you can always bind the "toggle run" keybinding or the movement mode menu, instead of using cycle movement mode

**Comment by Ramza13** [comment]
Is it worth adding an option to disable this and just use the previous behavior?

**Comment by esotericist** [comment]
> Is it worth adding an option to disable this and just use the previous behavior?

Could change the setting from a boolean to a three-way, though it'd need different framing

**Comment by Ramza13** [comment]
> > Is it worth adding an option to disable this and just use the previous behavior?
> 
> Could change the setting from a boolean to a three-way, though it'd need different framing

I'll probably make it four and allow you to set crouching as the movement style also.

**Comment by Unrepentant-Atheist** [comment]
I don't like it like it is now. Please make it an option, because I'm getting kinda annoyed by this: 
image

I could do that while walking not running before, now I have to switch to walk, then switch off again. C'mon....no, just no. Good idea, bad implementation.

**Comment by Ramza13** [comment]
I have a pr awaiting review right now to make it optional https://github.com/CleverRaven/Cataclysm-DDA/pull/36037

**Pull request merged**